tests/components/modbus/test_modbus_switch.py

```diff
@@ -184,42 +172,98 @@ async def test_register_switch(hass, regs, expected):
+    entity_id = f"{SWITCH_DOMAIN}.{switch_name}"
+    test_value = STATE_ON
+    config_switch = {CONF_NAME: switch_name, CONF_ADDRESS: 17}
+    with mock.patch(
+        "homeassistant.components.modbus.switch.ModbusSwitch.async_get_last_state"
+    ) as mock_get_last_state:
+        mock_get_last_state.return_value = State(entity_id, f"{test_value}")
+
+        await base_config_test(
+            hass,
+            config_switch,
+            switch_name,
+            SWITCH_DOMAIN,
+            CONF_SWITCHES,
+            None,
+            method_discovery=True,
+        )
+        assert hass.states.get(entity_id).state == test_value
+
+
+async def test_switch_service_turn(hass, caplog, mock_pymodbus):
+    """Run test for service turn_on/turn_off."""
+
+    entity_id1 = f"{SWITCH_DOMAIN}.switch1"
+    entity_id2 = f"{SWITCH_DOMAIN}.switch2"
+    config = {
+        MODBUS_DOMAIN: {
+            CONF_TYPE: "tcp",
+            CONF_HOST: "modbusTestHost",
+            CONF_PORT: 5501,
+            CONF_SWITCHES: [
+                {
+                    CONF_NAME: "switch1",
+                    CONF_ADDRESS: 17,
+                    CONF_WRITE_TYPE: CALL_TYPE_REGISTER_HOLDING,
+                },
+                {
+                    CONF_NAME: "switch2",
+                    CONF_ADDRESS: 17,
+                    CONF_WRITE_TYPE: CALL_TYPE_REGISTER_HOLDING,
+                    CONF_VERIFY: {},
+                },
+            ],
         },
-        switch_name,
-        SWITCH_DOMAIN,
-        CONF_SWITCHES,
-        None,
-        regs,
-        expected,
-        method_discovery=True,
-        scan_interval=5,
+    }
+    assert await async_setup_component(hass, MODBUS_DOMAIN, config) is True
+    await hass.async_block_till_done()
+    assert MODBUS_DOMAIN in hass.config.components
+
+    assert hass.states.get(entity_id1).state == STATE_OFF
+    await hass.services.async_call(
+        "switch", "turn_on", service_data={"entity_id": entity_id1}
     )
-    assert state == expected
+    await hass.async_block_till_done()
+    assert hass.states.get(entity_id1).state == STATE_ON
+    await hass.services.async_call(
+        "switch", "turn_off", service_data={"entity_id": entity_id1}
+    )
+    await hass.async_block_till_done()
+    assert hass.states.get(entity_id1).state == STATE_OFF
+
+    mock_pymodbus.read_holding_registers.return_value = ReadResult([0x00])
+    assert hass.states.get(entity_id2).state == STATE_OFF
+    await hass.services.async_call(
+        "switch", "turn_off", service_data={"entity_id": entity_id2}
+    )
+    await hass.async_block_till_done()
+    assert hass.states.get(entity_id2).state == STATE_OFF
+    mock_pymodbus.read_holding_registers.return_value = ReadResult([0x01])
+    await hass.services.async_call(
+        "switch", "turn_on", service_data={"entity_id": entity_id2}
+    )
+    await hass.async_block_till_done()
+    assert hass.states.get(entity_id2).state == STATE_ON
```

Comment:
Hi @janiversen. I pulled the latest dev and can't wrap my head around how swith.py ended up calling the `async_write_register`.

A swith.py ctor `_write_type` can only have `CALL_TYPE_REGISTER_HOLDING` and `CALL_TYPE_COIL` values.

The call in the `async_turn_on` function
```
        result = await self._hub.async_pymodbus_call(
            self._slave, self._address, self._command_on, self._write_type
        )
```
will end up reading the values instead of writing them.

Just thinking the test failure might be due to ^ that.

Reply:
yup see #50812.

good catch.

Reply:
be awere this PR is a draft, because I am still pursuing a couple of problems.